keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,462 @@
+          x-kubernetes-list-type: atomic
+          properties:
+            x:
+              type: string
+            y:
+              type: integer
+    ```
+    such that `{"someSet":[{"x":"abc"},{"x":"def","y":1}]}` and `{"someSet":[{"x":"def","y":1},{"x":"abc"}]}` are considered equally and not mutated.
+    
+- (Possible future extension) **Addition only/Deletion only**: Maps and arrays of list-type `map` and `set` can be addition-only and deletion-only by setting either `x-kubernetes-addition-only: true` or `x-kubernetes-deletion-only: true`, both mutually exclusive with `x-kubernetes-immutable` and `x-kubernetes-immutable-keys`.
+  
+  - For maps:
+    ```yaml
+    properties:
+      someMap:
+        type: object
+        x-kubernetes-addition-only: true # allowed with additionalProperties
+        additionalProperties:
+          type: string
+    ```
+    and similarly for `x-kubernetes-deletion-only: true`.
+      
+  - For arrays of list-type `map` we get the same behaviour, but enforce that the map-keys
+    are immutable as well (via CRD validation):
+      
+    ```yaml
+    properties:
+      someArray:
+        type: array
+        x-kubernetes-list-type: map
+        x-kubernetes-list-map-keys: ["name"]
+        x-kubernetes-addition-only: true # allowed with `x-kubernetes-list-type: map`
+        items:
+          properties:
+            name:
+              type: string
+              x-kubernetes-immutable: true # enforced
+            x:
+              type: string
+            y:
+              type: integer
+    ```
+    and similarly for `x-kubernetes-deletion-only: true`.
+      
+  - For arrays of list-type `set` get:
+  
+    ```yaml
+    properties:
+      someSet:
+        type: array
+        x-kubernetes-list-type: set
+        x-kubernetes-addition-only: true # allowed with `x-kubernetes-list-type: set`
+        items:
+          type: object
+          x-kubernetes-list-type: atomic
+          properties:
+            x:
+              type: string
+            y:
+              type: integer
+    ```
+    and similarly for `x-kubernetes-deletion-only: true`.
+    
+### Publishing
+
+The `x-kubernetes-immutable` and `x-kubernetes-immutable-keys` vendor extensions are published via `/openapi/v2` as is.
+
+### Suggested marker syntax
+
+In analogy to `+required`, `+optional` we propose to add a marker to kubebuilder's
+controller-gen named `+immutable`, which covers the whole subtree of the object 
+(compare field-selection section for extension).
+
+```
+// The name can not be changed after creation.
+// +immutable
+Name string
+
+// The list of containers can not change AT ALL after creation.
+// No single field in existing containers can be changed, added or deleted,
+// no new containers can be added, no existing container can be removed.
+// +immutable
+Containers []Containers
+```
+
+The `x-kubernetes-immutable-keys: true` vendor extension is expressed via `// +immutable=keys` in Golang types:
```

Comment:
My understanding was that you can have both immutable and immutable keys at the same time,

would you have to do:
```
// +immutable
// +immutable=keys
```

Reply:
We choose to not allow both at the same time. The former is stronger as the latter.

Note: an `// +marker` on a slice **does not** apply to the items, but on the slice itself. E.g. `// +nullable` allows to set `null` for a slice, but not `[null, null]` as value. Here `// +immutable`'s semantics is recursive, so `// +immutable=keys` is subsumed by `// +immutable` for the slice.